File: androidApp/src/main/java/com/cacd2/cacdgame/android/features/history/ui/HistoryFilterMenuView.kt
package com.cacd2.cacdgame.android.features.history.ui

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.material.Checkbox
import androidx.compose.material.CheckboxDefaults
import androidx.compose.material.DropdownMenu
import androidx.compose.material.DropdownMenuItem
import androidx.compose.material.MaterialTheme
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.PlatformTextStyle
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.DpOffset
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.cacd2.cacdgame.android.AppTheme
import com.cacd2.cacdgame.android.R
import com.cacd2.cacdgame.android.robotoFonts
import com.cacd2.cacdgame.model.GameChoice
import com.cacd2.cacdgame.model.HistoryFilterCriteria

/**
 * Created by francois.dabonot@cacd2.fr on 07/08/2023.
 */
@Composable
fun HistoryFilterMenuView(
    expanded: Boolean,
    filterCriteria: HistoryFilterCriteria,
    onUpdate: (HistoryFilterCriteria) -> Unit,
    onDismiss: (HistoryFilterCriteria) -> Unit
) {

    var design by remember { mutableStateOf(filterCriteria.designSelected) }
    var tech by remember { mutableStateOf(filterCriteria.techSelected) }
    var product by remember { mutableStateOf(filterCriteria.productSelected) }
    var mix by remember { mutableStateOf(filterCriteria.mixSelected) }
    var availableCategory by remember { mutableStateOf<List<GameChoice>>(emptyList()) }

    LaunchedEffect(Unit) {
        availableCategory = GameChoice.getAvailableHistoryChoice()
    }


    fun forwardData() {
        onUpdate(
            HistoryFilterCriteria(
                designSelected = design,
                techSelected = tech,
                productSelected = product,
                mixSelected = mix
            )
        )
    }

    fun dismissFilter() {
        onDismiss(
            HistoryFilterCriteria(
                designSelected = design,
                techSelected = tech,
                productSelected = product,
                mixSelected = mix
            )
        )
    }

    DropdownMenu(
        expanded = expanded,
        onDismissRequest = { dismissFilter() },
        offset = DpOffset(x = (-10).dp, y = (-30).dp)
    ) {
        if (availableCategory.contains(GameChoice.DESIGN)) {
            DropdownMenuItem(
                enabled = true,
                onClick = {
                    design = !design
                    forwardData()
                }
            ) {
                Row(
                    horizontalArrangement = Arrangement.spacedBy(12.dp),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Checkbox(
                        modifier = Modifier.padding(top = 2.dp),
                        checked = design,
                        onCheckedChange = null,
                        colors =
                        CheckboxDefaults.colors(
                            checkedColor = MaterialTheme.colors.primary,
                            disabledIndeterminateColor = MaterialTheme.colors.primary,
                            uncheckedColor = MaterialTheme.colors.primary
                        )
                    )
                    Text(
                        text = stringResource(id = R.string.selection_quiz_design),
                        modifier = Modifier,
                        style =
                        TextStyle(
                            fontFamily = robotoFonts,
                            fontSize = 15.sp,
                            platformStyle =
                            PlatformTextStyle(
                                includeFontPadding = false
                            )
                        ),
                        lineHeight = 18.sp
                    )
                }
            }
        }
        if (availableCategory.contains(GameChoice.TECH)) {
            DropdownMenuItem(
                enabled = true,
                onClick = {
                    tech = !tech
                    forwardData()
                }
            ) {
                Row(
                    horizontalArrangement = Arrangement.spacedBy(12.dp),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Checkbox(
                        modifier = Modifier.padding(top = 2.dp),
                        checked = tech,
                        onCheckedChange = null,
                        colors =
                        CheckboxDefaults.colors(
                            checkedColor = MaterialTheme.colors.primary,
                            disabledIndeterminateColor = MaterialTheme.colors.primary,
                            uncheckedColor = MaterialTheme.colors.primary
                        )
                    )
                    Text(
                        text = stringResource(id = R.string.selection_quiz_tech),
                        modifier = Modifier,
                        style =
                        TextStyle(
                            fontFamily = robotoFonts,
                            fontSize = 15.sp,
                            platformStyle =
                            PlatformTextStyle(
                                includeFontPadding = false
                            )
                        ),
                        lineHeight = 18.sp
                    )
                }
            }
        }
        if (availableCategory.contains(GameChoice.PRODUCT)) {
            DropdownMenuItem(
                enabled = true,
                onClick = {
                    product = !product
                    forwardData()
                }
            ) {
                Row(
                    horizontalArrangement = Arrangement.spacedBy(12.dp),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Checkbox(
                        modifier = Modifier.padding(top = 2.dp),
                        checked = product,
                        onCheckedChange = null,
                        colors =
                        CheckboxDefaults.colors(
                            checkedColor = MaterialTheme.colors.primary,
                            disabledIndeterminateColor = MaterialTheme.colors.primary,
                            uncheckedColor = MaterialTheme.colors.primary
                        )
                    )
                    Text(
                        text = stringResource(id = R.string.selection_quiz_produit),
                        modifier = Modifier,
                        style =
                        TextStyle(
                            fontFamily = robotoFonts,
                            fontSize = 15.sp,
                            platformStyle =
                            PlatformTextStyle(
                                includeFontPadding = false
                            )
                        ),
                        lineHeight = 18.sp
                    )
                }
            }
        }
        if (availableCategory.contains(GameChoice.ALL)) {
            DropdownMenuItem(
                enabled = true,
                onClick = {
                    mix = !mix
                    forwardData()
                }
            ) {
                Row(
                    horizontalArrangement = Arrangement.spacedBy(12.dp),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Checkbox(
                        modifier = Modifier.padding(top = 2.dp),
                        checked = mix,
                        onCheckedChange = null,
                        colors =
                        CheckboxDefaults.colors(
                            checkedColor = MaterialTheme.colors.primary,
                            disabledIndeterminateColor = MaterialTheme.colors.primary,
                            uncheckedColor = MaterialTheme.colors.primary
                        )
                    )
                    Text(
                        text = stringResource(id = R.string.selection_mix),
                        modifier = Modifier,
                        style =
                        TextStyle(
                            fontFamily = robotoFonts,
                            fontSize = 15.sp,
                            platformStyle =
                            PlatformTextStyle(
                                includeFontPadding = false
                            )
                        ),
                        lineHeight = 18.sp
                    )
                }
            }
        }
    }
}

@Preview
@Composable
fun HistoryFilterMenuPreview() {
    AppTheme {
        HistoryFilterMenuView(
            expanded = true,
            filterCriteria =
            HistoryFilterCriteria(
                designSelected = true,
                techSelected = true,
                productSelected = true,
                mixSelected = false
            ),
            onUpdate = {
                print(it)
            },
            onDismiss = {
                print(it)
            }
        )
    }
}

File: iosApp/iosApp/Features/History/UI/HistoryFilterMenuView.swift
//
//  HistoryFilterMenuView.swift
//  CACDGAME
//
//  Created by François Dabonot on 07/08/2023.
//  Copyright © 2023 CACD2. All rights reserved.
//

import SwiftUI
import Shared

extension HistoryFilterCriteria: ObservableObject {
}

struct HistoryFilterMenuView: View {

    @Binding var filterCriteria: HistoryFilterCriteria

    @State private var design: Bool
    @State private var tech: Bool
    @State private var product: Bool
    @State private var mix: Bool
    @State private var availableCategory: [GameChoice] = []

    init(filterCriteria: Binding<HistoryFilterCriteria>) {
        self._filterCriteria = filterCriteria
        self.design = filterCriteria.designSelected.wrappedValue
        self.tech = filterCriteria.techSelected.wrappedValue
        self.product = filterCriteria.productSelected.wrappedValue
        self.mix = filterCriteria.mixSelected.wrappedValue
    }

    private func updateCriteria() {
        filterCriteria = HistoryFilterCriteria(designSelected: design,
                                               techSelected: tech,
                                               productSelected: product,
                                               mixSelected: mix)
    }

    var body: some View {
        VStack(spacing: 12) {
            if availableCategory.contains(.design) {
                Toggle(R.string.localizable.selection_quiz_design.string, isOn: $design)
                    .font(R.font.robotoRegular.font(size: 15))
            }
            if availableCategory.contains(.tech) {
                Toggle(R.string.localizable.selection_quiz_tech.string, isOn: $tech)
                    .font(R.font.robotoRegular.font(size: 15))
            }
            if availableCategory.contains(.product) {
                Toggle(R.string.localizable.selection_quiz_produit.string, isOn: $product)
                    .font(R.font.robotoRegular.font(size: 15))
            }
            if availableCategory.contains(.all) {
                Toggle(R.string.localizable.selection_mix.string, isOn: $mix)
                    .font(R.font.robotoRegular.font(size: 15))
            }
        }
        .padding()
        .onChange(of: design) { _ in
            updateCriteria()
        }.onChange(of: tech) { _ in
            updateCriteria()
        }.onChange(of: product) { _ in
            updateCriteria()
        }.onChange(of: mix) { _ in
            updateCriteria()
        }
        .task {
            do {
                availableCategory = try await GameChoice.companion.getAvailableHistoryChoice()
            } catch {
                AppLogger.shared.e(messageString: "Can't retrieve available category")
            }
        }
    }
}

#Preview {
    HistoryFilterMenuView(filterCriteria: .constant(.init(designSelected: true,
                                                          techSelected: true,
                                                          productSelected: true,
                                                          mixSelected: true)))
}
